### PR TITLE
Notation

### PR DESCRIPTION
**Pull request by battermann**
There are two tests/examples for the ABC notation.

One produces this output:

screenshot 2018-11-15 at 22 08 26

The other one produces this output:

screenshot 2018-11-15 at 22 10 36

I have tried to find a good balance between

* usability / flexability
* making illegal state unrepresentable / type safety

I think being extreme in either of those directions won't work well in this particular context. But I'm open for any suggestions to improve things.

Note, that some implementations are not complete. This is still WIP.

### CODE REVIEW

**Comment by duncanmalashock** [comment]
Wow. I'm very impressed!

**Review by duncanmalashock — COMMENTED** [review]
When will this PR be ready to merge?

**Comment by battermann** [comment]
> When will this PR be ready to merge?

Must haves:

* More Tuplet constructors
* ABC: specify clef for staffs
* Add more tests
* Capabilities for `Notation Pitch -> ABC` (current tests use `Notation PitchSpelling`)
  * Therefore we need some more constructors for pitches 
  * And we need `sequence` and `traverse` for `Notation`: `(a -> Result e b) -> Notation a -> Result e (Notation b)` etc...

Nice to have:

* ABC support for all the note attributes
* Support for displaying accidentals according to key signature and previous accidentals, e.g. accidentals should not have to be repeated within one bar (this will be a little more complex, IMO)
* Functions to validate a `Notation a`
* Function to divide a notation into bars if no bars are specified
* ...

IMO there are no show stoppers.

These are just some ideas, let me know what you think.

**Comment by duncanmalashock** [comment]
Those all sound like good features. Is it possible to merge the open PR and
start doing these one at a time? I am starting to get uncomfortable because
I haven’t really been able to collaborate with you for a few weeks.

File: src/MusicTheory/Duration.elm [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    , oneHundredTwentyEighthNote
+    , quarterNote
+    , sixteenthNote
+    , sixtyFourthNote
+    , thirtySecondNote
+    , toList
+    , toRational
+    , tripleDotted
+    , wholeNote
+    , zero
```
Comment (duncanmalashock) [inline src/MusicTheory/Duration.elm]: does this need to be exposed?

File: src/Libs/Ratio.elm [hunk 2]
```diff
@@ -0,0 +1,323 @@
+module Libs.Ratio exposing
```
Comment (duncanmalashock) [inline src/Libs/Ratio.elm:1]: Is there a way we could load this as a dependency?

File: src/MusicTheory/Notation.elm [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+    , Voice(..)
+    , VoiceAttribute(..)
+    , VoiceElement(..)
+    , bar
+    , bass
+    , beam
+    , chord
+    , chordNote
+    , doubleLine
+    , elMap
```
Comment (duncanmalashock) [inline src/MusicTheory/Notation.elm]: is `elMap` just an implementation detail of `voiceMap`, or will client code use it also? If not, then maybe we don't expose it?

**Review by duncanmalashock — APPROVED** [review]

File: src/MusicTheory/Duration.elm (hunk 1 above)
Comment (battermann) [reply src/MusicTheory/Duration.elm]: No, I'll remove it.

File: src/Libs/Ratio.elm (hunk 2 above)
Comment (battermann) [reply src/Libs/Ratio.elm:1]: Unfortunately the [ratio lib](https://package.elm-lang.org/packages/elm-community/ratio/latest/Ratio) has not been updated to 0.19. So we can't ATM.

File: src/MusicTheory/Notation.elm (hunk 3 above)
Comment (battermann) [reply src/MusicTheory/Notation.elm]: Yes, that's a detail that does not need to be exposed. There is also `staffMap` and `voiceMap`, that are used for `map`. I think it's ok to leave them exposed for now.

**Pull request merged**